example-plugins/inj-tyfam-plugin/src/Plugin/InjTyFam.hs

```diff
@@ -0,0 +1,242 @@
+
+            -- contradiction!
+            | Unify.typesCantMatch [(lty, rty)] ->
+            pure Nothing
+
+            -- emit new constraint and update the RHS argument
+            | otherwise -> do
+            new <- impliedGivenEq ct lty rty
+            go (lty : acc1) (new : acc2) args
+
+        [] -> pure $ Just (reverse acc1, acc2)
+
+    -- replace the equality with one with updated RHS arguments
+    ok :: TcType -> TyCon -> ([Xi], [Ct]) -> M R
+    ok lty tc (rtys', news)
+        | null news = skip   -- we made no changes
+        | otherwise = do
+        new <- impliedGivenEq ct lty rty'
+        pure $ TcRnTypes.TcPluginOk [(ev, ct)] (new:news)
+      where
+        rty' :: TcType
+        rty' = Type.mkTyConApp tc rtys'
+
+        ev :: EvTerm
+        ev = TcRnTypes.ctEvTerm $ TcRnTypes.cc_ev ct
+
+-- | Emit a new Given equality constraint implied by another Given
+-- equality constraint
+
+impliedGivenEq :: Ct -> TcType -> TcType -> M Ct
+impliedGivenEq ct lty rty = do
+    let
+      new_co :: Coercion
+      new_co = Coercion.mkUnivCo
+        (TyCoRep.PluginProv "inj-tyfam-plugin") TyCon.Nominal lty rty
+
+    -- TODO how to incorporate @ctEvId ct@? (see #15248 on GitLab ghc)
+    new_ev <- TcPluginM.newGiven
+      (TcRnTypes.bumpCtLocDepth (TcRnTypes.ctLoc ct))  -- TODO don't bump?
+      (Type.mkPrimEqPredRole TyCon.Nominal lty rty)
+      (CoreSyn.Coercion new_co)
+
+    pure $ TcRnTypes.mkNonCanonical new_ev
+
+-----
+
+-- | See 'mkFskEnv'
+
+type FskEnv = VarEnv (TyCon, [Xi])
+
+-- | An map created from the @FunEq@s
+--
+-- NOTE not necessarily idempotent
+
+mkFskEnv :: [Ct] -> FskEnv
+mkFskEnv cts =
+    VarEnv.mkVarEnv
+    [ (cc_fsk, (cc_fun, cc_tyargs))
+    | TcRnTypes.CFunEqCan{..} <- cts
+    ]
+
+-- | Substitute via the 'FskEnv' once
+
+unfsk :: FskEnv -> TcType -> TcType
+unfsk fskEnv t = fromMaybe t $ do
+    v <- Type.getTyVar_maybe t
+    uncurry Type.mkTyConApp <$> VarEnv.lookupVarEnv fskEnv v
+
+-----
+
+-- | A pair of arguments involved in an actionable equality constraint
+
+data Triple = MkTriple Xi Xi Bool
+      -- ^ the LHS argument, the RHS argument, and whether this
+      -- argument is included in the family's injectivity annotation
+
+-- | See 'Triple'
+
+splitter :: FskEnv -> TcType -> TcType -> Maybe (TyCon, [Triple])
+splitter fskEnv = \lty rty -> do
+    (ltc, ltys) <- prj lty
+    (rtc, rtys) <- prj rty
+
+    guard $ ltc == rtc
+
+    case TyCon.tyConInjectivityInfo ltc of
+      TyCon.NotInjective -> Nothing
+      TyCon.Injective is ->
+          -- INVARIANT: @True == or is@
+          Just (ltc, go is ltys rtys)
+
+  where
+    prj :: TcType -> Maybe (TyCon, [Xi])
+    prj = Type.splitTyConApp_maybe . unfsk fskEnv
+
+    go :: [Bool] -> [Xi] -> [Xi] -> [Triple]
+    go []     []         []         = []
+    go (i:is) (lty:ltys) (rty:rtys) =
+        MkTriple lty rty i : go is ltys rtys
+    go _      _          _          = error "impossible"
```

Comment:
is it really impossible? The documentation for `cc_tyargs` says the list could be under-saturated, in which case there would be fewer arguments than booleans. Also, if the user accidentally writes a type equality with the wrong arity, e.g. `F a1 a2 ~ F b1 b2 b3`, are we guaranteed that a type error will be raised before the execution reaches this line?

Reply:
I believe so. I'd like the plugin to fail if I'm wrong. A length mismatch there will be so problematic that GHC will set that constraint aside, I think.

Regarding unsaturatedness -- I believe that since we're looking at the left- and right-sides of an `~` constraint, GHC will have rejected an under-saturates tyfam applications in such a context.